src/preact/component/intersection-observer.js

```diff
@@ -22,10 +23,40 @@ export function useIntersectionObserver(callback) {
       if (!node) {
         return;
       }
-      unobserveRef.current = observeIntersections(node, callback);
+
+      unobserveRef.current = observeIntersections(node, callback, ioOptions);
     },
-    [callback]
+    [callback, ioOptions]
```

Comment:
If the `ioOptions` change, would this cause the `InObs` to be recreated?

Reply:
Yes, that's true, the old one would be unobserved and a new one would be created.  This means that "inline configuration" would not work well (eg. `useIntersectionObserver(() => {...}, { threshold: 1 });` and instead a constant reference is needed (via `useMemo` or just a top-level `const`).

I did consider destructuring the `ioOptions` instead, like:
```suggestion
    [callback, ioOptions.threshold, ioOptions.needsRootBounds, ioOptions.rootMargin]
```
... the only problem would be if a new field was ever added to the `ioOptions` type, it would be hard to catch this error.  So then I considered something like:
```suggestion
    [callback, ...Object.values(ioOptions)]
```
But that could kinda break the rules of hooks.  So I figured the consistent reference via `useMemo` was probably the least problematic.

What do you think?

Reply:
Yeah, I see the need for the dependency. I would suggest returning a cleanup function to disconnect the previous `InOb`.

Reply:
Proper cleanup is happening (not in this PR, it was already existing).  
The "cleanup" for particular hook, though, looks different than the standard `useEffect` cleanup, but it's in there:

https://github.com/ampproject/amphtml/blob/a2fbde35a1140a2255801e257a63bf2c8a2091f2/src/preact/component/intersection-observer.js#L17-L21